perceval/algorithm/tomography/tomography_mle.py

```diff
@@ -0,0 +1,400 @@
+        # uses measurements to reconstruct quantum maps (state or process) using MLE for Quantum Tomography
+        # ref: https://doi.org/10.48550/arXiv.1609.07881
+        #
+        # param init_guess_quantum_map: an initial guess ofr the quantum map
+
+        guess_quantum_map = copy.deepcopy(init_guess_quantum_map)
+        init_quantum_map = copy.deepcopy(init_guess_quantum_map)
+
+        theta = 1
+        ith_learn_rate = self.init_learn_rate  # initializing the learning rate of the algorithm
+
+        log_f_guess_quantum_map = self._log_likelihood_func(guess_quantum_map)
+        grad_log_f_guess_quantum_map = self._grad_log_likelihood_func(guess_quantum_map)
+
+        for i in range(self.max_iterations):
+            ith_quantum_map = self._proj(guess_quantum_map - ith_learn_rate * grad_log_f_guess_quantum_map)
+
+            delta_i = ith_quantum_map - guess_quantum_map  # difference between current and target
+
+            log_f_ith_quantum_map = self._log_likelihood_func(ith_quantum_map)
+
+            frob_prod_grad_log_f_delta = self._frobenius_inner_product(grad_log_f_guess_quantum_map, delta_i)
+
+            norm_delta_i = (1 / (2 * ith_learn_rate)) * np.linalg.norm(delta_i, ord='fro') ** 2
+
+
+            while log_f_ith_quantum_map > (log_f_guess_quantum_map + frob_prod_grad_log_f_delta + norm_delta_i):
+                ith_learn_rate *= self.decelerate_factor
+                ith_quantum_map = TomographyMLE._proj(guess_quantum_map - ith_learn_rate * grad_log_f_guess_quantum_map)
+                delta_i = ith_quantum_map - guess_quantum_map
+
+            delta_i_hat = ith_quantum_map - init_quantum_map
+
+            if TomographyMLE._frobenius_inner_product(delta_i, delta_i_hat) < 0:
+                # Restart by re-initializing guess maps
+                ith_quantum_map, guess_quantum_map, theta = init_quantum_map, init_quantum_map, 1
+            else:
+                # Accelerate the algorithm in the next iteration
+                theta, guess_quantum_map = (1 + np.sqrt(1 + 4 * theta ** 2)) / 2, ith_quantum_map + delta_i_hat * (theta - 1) / (
+                            (1 + np.sqrt(1 + 4 * theta ** 2)) / 2)
+
+            if np.abs(log_f_ith_quantum_map - self._log_likelihood_func(init_quantum_map)) < self.convergence_precision:
+                break
+
+            init_quantum_map = ith_quantum_map
+
+        return ith_quantum_map
+
+
+
+class StateTomographyMLE(TomographyMLE):
+    def __init__(self, operator_processor, **kwargs):
+        super().__init__(operator_processor, **kwargs)
+        self._f = self._povm_data()
+        self._guess_density_matrix = np.eye(2 ** self._nqubit) / (2 ** self._nqubit)
+
+    def _povm_data(self) -> list:
+        # Performing a POVM (positive operator value measure) on the quantum processor
+        # in the informationally complete Pauli basis, i.e. choosing all the eigenvectors of the PAuli operators.
+        # They are |0>,|1>,|+>,|->,|i+>,|i->
+
+        f = []
```

Comment:
I would move the f declaration down (just above the loop).
As a rule, always declare variable as later as possible.